client/src/hooks/useSession.ts

```diff
@@ -37,28 +37,33 @@ export const useSession = () => {
   // Until the user initiates a login or logout, we don't want to create or
   // destroy a session
   const [canAlterSession, setCanAlterSession] = useState(false);
+  // isAuthenticated does not change immediately after login/logout, so we
+  // need to track the intent to prevent accidental session creation/deletion
+  const [toLogout, setToLogout] = useState(false);
 
   const login = async () => {
     await loginAuth();
     setCanAlterSession(true);
+    setToLogout(false);
```

Comment:
Here's something I don't understand. Is it safe to expect, the `useEffect` callback will not run after the `setCanAlterSession`, but before `setToLogout`?

It doesn't seem to be doing that, but if it would, then the wrong action on session would be run.

Reply:
Good question. I think it's fine and I'll try to explain why.

How I understand the interaction between state, effect and renders is that `useEffect` runs after each render and `setState` signals to React that it may need to render.  However, it doesn't immediately trigger a render, it lets the rest of the component logic resolve first.  That means that both `setCanAlterSession` and `setToLogout` will resolve and alter their respective states. At that point the page can render and only then will the `useEffect` fire.

Reply:
That makes sense, I wasn't sure if that's the expected behavior or just being lucky when testing it. Thanks :rocket